File: patterns/strategy/strategy.js
// http://www.dofactory.com/javascript-strategy-pattern.aspx
//
// Note since this is JavaScript and there are no interfaces (see duck typing)
// there is no Strategy Participant in this example

// Context
var Shipping = function() {
  this.company = "";
};

// Context
Shipping.prototype = {
  setStrategy: function(company) {
    this.company = company;
  },
  calculate: function(package) {
    return this.company.calculate(package);
  }
};

// ConcreteStrategy
var UPS = function() {
  this.calculate = function(package) {
    // calculations...
    return "$45.95";
  }
};

// ConcreteStrategy
var USPS = function() {
  this.calculate = function(package) {
    // calculations...
    return "$39.40";
  }
};

// ConcreteStrategy
var Fedex = function() {
  this.calculate = function(package) {
    // calculations...
    return "$43.20";
  }
};

// log helper
var log = (function() {
  var log = "";
  return {
    add: function(msg) { log += msg + "\n"; },
    show: function() { console.log(log); log = ""; }
  }
})();

function run() {
  var package = { from: "76712", to: "10012", weigth: "lkg" };

  // the 3 strategies

  var ups = new UPS();
  var usps = new USPS();
  var fedex = new Fedex();

  var shipping = new Shipping();
  shipping.setStrategy(ups);
  log.add("UPS Strategy: " + shipping.calculate(package));

  shipping.setStrategy(usps);
  log.add("USPS Strategy: " + shipping.calculate(package));

  shipping.setStrategy(fedex);
  log.add("Fedex Strategy: " + shipping.calculate(package));

  log.show();
}

run();

File: patterns/strategy/strategy.cs
using System;

namespace StrategyPattern
{
    //context
    public abstract class Company
    {
        public abstract double Calculate(Package package);
    }

    //context
    public class Shipping
    {
        private Company company;

        public void SetStrategy(Company company)
        {
            this.company = company;
        }

        public double Calculate(Package package)
        {
            return company.Calculate(package);
        }
    }

    //dependent 
    public class Package
    {
        public int From { get; set; }
        public int To { get; set; }
        public string Weight { get; set; }
    }

    //concreteStrategy
    public class UPS : Company
    {
        public override double Calculate(Package package)
        {
            // calculations...
            return 45.95;
        }
    }

    //concreteStrategy
    public class USPS : Company
    {
        public override double Calculate(Package package)
        {
            // calculations...
            return 39.40;
        }
    }

    //concreteStrategy
    public class Fedex : Company
    {
        public override double Calculate(Package package)
        {
            // calculations...
            return 43.20;
        }
    }

    //client Participant
    internal class Program
    {
        static void Main(string[] args)
        {
            Package package = new Package() 
                            { From  = 76712, To = 10012, Weight = "1kg"};

            // the 3 strategies
            UPS ups = new UPS();
            USPS usps = new USPS();
            Fedex fedex = new Fedex();

            Shipping shipping = new Shipping();
            shipping.SetStrategy(ups);
            Console.WriteLine("UPS Strategy: " + shipping.Calculate(package));

            shipping.SetStrategy(usps);
            Console.WriteLine("USPS Strategy: " + shipping.Calculate(package));

            shipping.SetStrategy(fedex);
            Console.WriteLine("Fedex Strategy: " + shipping.Calculate(package));
            
        }
    }
}
